Source/DafnyCompilers/Interp.dfy

```diff
@@ -93,13 +98,83 @@ module Interp {
     }
   }
 
-  predicate method WellFormedValue1(v: V.T)  {
+  // TODO: rewrite as a shallow predicate applied through ``v.All``?
+  predicate method WellFormedEqValue(v: V.T)
+  // This predicate gives the constrainst we need to be able to *define* our equivalence relation
+  // over values and actually *use* this relation to prove equivalence properties between expressions.
+  // 
+  // The difficult point is linked to closures: when we apply transformations to the code, we often
+  // apply them in a deep manner to the expressions (i.e., to all the subexpressions of an expression).
+  // The problem is that it can have an effect on the closure values generated through the execution
+  // by modifying their bodies, leading to discrepancies in the execution of the code (imagine the case
+  // where we use closures as keys inside of maps).
+  // The good news is that when those cases happen, we actually try to use an equality over values
+  // which don't have a decidable equality: we solve the problem by forcing some subvalues to have
+  // a decidable equality.
+  {
+    match v {
+      case Bool(b) => true
+      case Char(c) => true
+      case Int(i) => true
+      case Real(r) => true
+      case BigOrdinal(o) => true
+      case BitVector(width, val) =>
+        0 <= val < Math.IntPow(2, width)
+      case Map(m) =>
+        && (forall x | x in m :: HasEqValue(x))
+        && (forall x | x in m :: WellFormedEqValue(x) && WellFormedEqValue(m[x]))
+      case Multiset(ms) =>
+        && HasEqValue(v)
+        && (forall x | x in ms :: WellFormedEqValue(x))
+      case Seq(sq) =>
+        && (forall x | x in sq :: WellFormedEqValue(x))
+      case Set(st) =>
+        && HasEqValue(v)
+        && (forall x | x in st :: WellFormedEqValue(x))
+      case Closure(ctx, vars, body) =>
+        // TODO: is that enough?
+        && (forall x | x in ctx.Values :: WellFormedEqValue(x))
+    }
+  }
+
+  // TODO: rename to ValueHasEq
+  predicate method HasEqValue(v: V.T)
+  // Return true if the value supports a decidale equality.
+  //
+  // Note that this is a bit subtle for collections: any empty collection supports a decidable
+  // equality, but non-empty collections support a decidable equality only if their elements
+  // support one.
```

Comment:
It's a bit surprising that decidable equality is a property of values, rather than of their types… but I guess we're doing it this way because we don't currently have types on values?